cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetContext.java

```diff
@@ -135,15 +137,16 @@ public DynamicDatasetContext(TransactionContext context, DatasetFramework datase
       dsArguments.putAll(arguments);
     }
 
+    String datasetKey = String.format("%s.%s", name, dsArguments.hashCode());
```

Comment:
I think it would be cleaner to have a dataset key class with the 2 fields, which can be internal to this class, and use that as the map key instead of formatting the string.